### PR TITLE
[Tensor] TP Linear 1D row

### PR DESCRIPTION
**Pull request by Wesley-Jzy**
Implement a Linear OP with Tensor Parallelism using 1D row splitting.

### CODE REVIEW

File: tests/test_tensor/test_linear_tp.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    assert torch.allclose(A, B, rtol=1e-3, atol=1e-1) == True
+
+def run_linear_tp1d_row_test():
+    device = get_current_device()
+    dtype = torch.float32
+    DEPTH = 4
+    in_features = 4
+    out_features = 5
+
+    i = gpc.get_local_rank(ParallelMode.PARALLEL_1D)
```
Comment (feifeibear) [inline tests/test_tensor/test_linear_tp.py:29]: use a meaningful var name. For example tp_local_rank =

File: tests/test_tensor/test_linear_tp.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    W = W.clone()
+    W.requires_grad = True
+
+    B_shape = (out_features)
+    B_master = torch.randn(B_shape, dtype=dtype, device=device)
+    dist.broadcast(B_master, src=0)
+    B = B_master.clone()
+    B.requires_grad = True
+
+    # replace the torch nn.Parameters with ShardedTensor
```
Comment (feifeibear) [inline tests/test_tensor/test_linear_tp.py:52]: with ColoTensor

File: tests/test_tensor/test_linear_tp.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    C_master.backward(grad_master)
+
+    W_grad = W_master.grad
+    W_grad = torch.chunk(W_grad, DEPTH, dim=-1)[i]
+    check_equal(W_grad, layer.weight.grad)
+
+    B_grad = B_master.grad
+    check_equal(B_grad, layer.bias.grad)
+
+    print('linear_row backward: pass')
```
Comment (feifeibear) [inline tests/test_tensor/test_linear_tp.py:92]: remove this line

File: tests/test_tensor/test_linear_tp.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    layer.weight.requires_grad = True
+    layer.bias.requires_grad = True
+    out = layer(A)
+
+    A_master = A_master.clone()
+    A_master.requires_grad = True
+    W_master = W_master.clone()
+    W_master.requires_grad = True
+    B_master = B_master.clone()
+    B_master.requires_grad = True
```
Comment (feifeibear) [inline tests/test_tensor/test_linear_tp.py:69]: why clone()?

File: tests/test_tensor/test_linear_tp.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    dist.broadcast(A_master, src=0)
+    A = A_master.clone()
+    A.requires_grad = True
+
+    W_shape = (out_features, in_features)
+    W_master = torch.randn(W_shape, dtype=dtype, device=device)
+    dist.broadcast(W_master, src=0)
+    W = torch.chunk(W_master, DEPTH, dim=-1)[i]
+    W = W.clone()
+    W.requires_grad = True
```
Comment (feifeibear) [inline tests/test_tensor/test_linear_tp.py:44]: The code snippet makes each process has the same tensor.
You can move these to a function.
The code is a little bit redundant

File: tests/test_tensor/test_linear_tp.py (hunk 1 above)
Comment (Wesley-Jzy) [reply tests/test_tensor/test_linear_tp.py:29]: Done

File: tests/test_tensor/test_linear_tp.py (hunk 2 above)
Comment (Wesley-Jzy) [reply tests/test_tensor/test_linear_tp.py:52]: Done.

File: tests/test_tensor/test_linear_tp.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    layer.bias.requires_grad = True
+    out = layer(A)
+
+    A_master = A_master.clone()
+    A_master.requires_grad = True
+    W_master = W_master.clone()
+    W_master.requires_grad = True
+    B_master = B_master.clone()
+    B_master.requires_grad = True
+    C_master = torch.matmul(A_master, W_master.transpose(0, 1)) + B_master
```
Comment (feifeibear) [inline tests/test_tensor/test_linear_tp.py:70]: why not use Linear?

File: colossalai/tensor/_ops/linear.py [hunk 7]
```diff
@@ -19,12 +20,31 @@ def colo_linear(types, args, kwargs, pg):
...
-        return torch.nn.functional.linear(input_tensor, weight.torch_tensor(), bias)
+        if weight.shard_spec == None:
+            return torch.nn.functional.linear(input_tensor, weight.torch_tensor(), bias)
+        elif weight.shard_spec == '1Drow':
+            """
+            Input:S[1] x Weight:S[0] = Output:P
+            All-Reduce(Output) + bias = res
+            """
+            # Input:S[1]
+            input_per_partition = split_forward_gather_backward(input_tensor, ParallelMode.PARALLEL_1D, dim=-1)
```
Comment (FrankLeeeee) [inline colossalai/tensor/_ops/linear.py:37]: What if the input has already been split by the row linear?

File: tests/test_tensor/test_linear_tp.py (hunk 3 above)
Comment (Wesley-Jzy) [reply tests/test_tensor/test_linear_tp.py:92]: Done.

File: tests/test_tensor/test_linear_tp.py (hunk 4 above)
Comment (Wesley-Jzy) [reply tests/test_tensor/test_linear_tp.py:69]: Just ensure the tesnors' clarity. I will remove them.

File: colossalai/tensor/_ops/linear.py (hunk 7 above)
Comment (Wesley-Jzy) [reply colossalai/tensor/_ops/linear.py:37]: Do you mean the input has already been split by 1D-Col split?

Comment (Wesley-Jzy) [reply colossalai/tensor/_ops/linear.py:37]: This pr is just working for complete input. The other situation will also be considered in the future.

Comment (FrankLeeeee) [reply colossalai/tensor/_ops/linear.py:37]: Ok, cool.

File: tests/test_tensor/test_linear_tp.py (hunk 6 above)
Comment (Wesley-Jzy) [reply tests/test_tensor/test_linear_tp.py:70]: Done.

File: tests/test_tensor/test_linear_tp.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+
+def run_dist(rank, world_size, port):
+    config = dict(parallel=dict(tensor=dict(mode="1d", size=world_size),))
+    colossalai.launch(config=config, rank=rank, world_size=world_size, host='localhost', port=port, backend='nccl')
+    run_linear_tp1d_row_test()
+
+
+@pytest.mark.dist
+@pytest.mark.parametrize("world_size", [4])
```
[the remark this answers is not in the document]
Comment (Wesley-Jzy) [reply tests/test_tensor/test_linear_tp.py:102]: Done.

File: tests/test_tensor/test_linear_tp.py (hunk 5 above)
Comment (Wesley-Jzy) [reply tests/test_tensor/test_linear_tp.py:44]: Done.

File: colossalai/tensor/_ops/linear.py (hunk 7 above)
Comment (feifeibear) [reply colossalai/tensor/_ops/linear.py:37]: we can check the input shape here. Make it works for our supported cases!
@Wesley-Jzy

**Review by feifeibear — APPROVED** [review]

File: colossalai/tensor/_ops/linear.py [hunk 9]
```diff
@@ -19,12 +20,31 @@ def colo_linear(types, args, kwargs, pg):
...
     if isinstance(bias, ColoTensor):
         bias = bias.torch_tensor()
 
     # Add communication logic before and after linear call.
     if isinstance(weight, ColoTensor):
-        return torch.nn.functional.linear(input_tensor, weight.torch_tensor(), bias)
+        if weight.shard_spec == None:
+            return torch.nn.functional.linear(input_tensor, weight.torch_tensor(), bias)
+        elif weight.shard_spec == '1Drow':
+            """
```
Comment (feifeibear) [inline colossalai/tensor/_ops/linear.py:32]: use. #

**Pull request merged**